Adds RATELIMIT_HASH_ALGORITHM config

slick666 opened the pull request:
This is just a simple update to the library that changes the default hashing off of md5. Md5 has been proven to be insecure, subject to collision attacks [[Wang](https://www.rfc-editor.org/rfc/rfc9155#Wang)]. In 2011, [[RFC6151](https://www.rfc-editor.org/rfc/rfc9155#RFC6151)] detailed the security considerations, including collision attacks for MD5. NIST formally deprecated use of SHA-1 in 2011 [[NISTSP800-131A-R2](https://www.rfc-editor.org/rfc/rfc9155#NISTSP800-131A-R2)]

This updates the hashing algorithm to an implementation without known collision issues and giving end users the option to configure the hashing function used for signature generation as their needs require. Implementation of the config check was modeled off of the decorator's check for `RATELIMIT_EXCEPTION_CLASS`

Finally, one small tweak of the tox file was needed to get tests to run locally.

benjaoming commented:
Can you add some motivation for the change?

When the default was already `md5`, isn't it better to leave it like that? md5 is faster?

slick666 commented:
Added the motivation to the main description. I'm posting my performance evaluation here for discussion. I did a simple test on my local machine to help understand potential impact. I took a simple walk through available hashlib algorithms and run a `timeit()` call on each with a random bytestring of 16 bytes. 16 bytes seems close to what we would typically see in production but I guessing here. I also moved the random generation to the setup phase to eliminate measuring that. Here is what I found.

#### Code
```
>>> for algo in [ x for x in dir(hashlib) if x.startswith('sha') or x == 'md5' ]:
...     t = timeit.Timer(f"hashlib.{algo}(sample).hexdigest()", "import hashlib; import random; sample=bytearray(random.getrandbits(8) for _ in range(16))")
...     print(f"{algo}:\t\t{t.timeit()}")
...
md5:		0.6161478290014202
sha1:		0.601631659999839
sha224:		0.7151631529995939
sha256:		0.6906373469973914
sha384:		0.7764537720067892
sha3_224:	1.0566239000036148
sha3_256:	1.018879391005612
sha3_384:	1.0503311959910206
sha3_512:	1.0615530710056191
sha512:		0.7907670210115612
```

#### Calculated Difference
|  Algorithm  | Percent Difference |
|------------|--------------------|
| md5:           | 0.00%                      |
| sha1:           | 2.36%                      |
| sha224:      | -16.07%                   |  
| sha256:      | -12.09%                  |
| sha384:      | -26.02%                  |
| sha3_224:  | -71.49%                   |
| sha3_256:  | -65.36%                  |
| sha3_384:  | -70.47%                  |
| sha3_512:   | -72.29%                  |
| sha512:        | -28.34%                 |

Based on this I might change the default from sha512 to sha256 since both are considered secure.

benjaoming commented:
Thanks for taking time to provide a really solid and educational motivation for the change :+1: 

So, just to be clear here:

1. We are talking about an attacker using knowledge about django-ratelimit to bypass the ratelimiting by means of an md5 collision attack?
2. Some systems that are security certified have md5 banned, right? IIRC, Python doesn't have md5 when running in "FIPS" mode.

jsocol commented:
Agreed! I really appreciate the data and the references here—and of course the actual update. It does seem like sha256 is a reasonable compromise of performance and security as a default, and if folks have the need to push that up to sha512 or one of the sha3 algorithms, they can.

slick666 commented:
Hey all, had to put this down for a couple weeks but I had a chance to pick this up and do some more poking at it. I included the code I used to check in case anyone in interesting in repeating what I did here. I included it here since it's just a one-off. Results are in the comments but the TLDR is this.

1. Shifting from md5 to sha256 doesn't appear to impact the overall performance of _make_cache_key(...). In fact running through all the algorithms shows very little change.
2. The impact of adding the additional check and import in _make_cache_key() had an immeasurable effect
3. There were other faster algorithms but based on these results I suggest we stick with sha256 because of it's NIST acceptance and call it good.

LMK if anything else is needed to merge, Thanks again for everyone's feedback

```
import hashlib
import statistics
import timeit

from django.test import TestCase


class PerformanceTestCore(TestCase):
    """
    Performance tests used to help determine the viability of different algorithms
    """
    ALGO_IGNORED_LIST = ['shake_128', 'shake_256']
    CACHE_KEY_SIZE = 16

    def test_algos(self):
        """
        Simple benchmark of algorithms

        Local Results:
            Baseline time to execute md5: 0.68720

            Time to execute blake2b:      0.41561 | % diff 0.6535%
            Time to execute blake2s:      0.29639 | % diff 1.3186%
            Time to execute md5:          0.65031 | % diff 0.0567%
            Time to execute sha1:         0.67619 | % diff 0.0163%
            Time to execute sha224:       0.75563 | % diff -0.0906%
            Time to execute sha256:       0.75148 | % diff -0.0855%
            Time to execute sha384:       0.82774 | % diff -0.1698%
            Time to execute sha3_224:     1.11615 | % diff -0.3843%
            Time to execute sha3_256:     1.13023 | % diff -0.3920%
            Time to execute sha3_384:     1.13423 | % diff -0.3941%
            Time to execute sha3_512:     1.16434 | % diff -0.4098%
            Time to execute sha512:       0.85234 | % diff -0.1937%
        """
        LENGTH = 16  # The Key length in bytes, this represents the average function name length

        base = timeit.Timer(
            f"hashlib.md5(sample).hexdigest()",
            f"import hashlib; import random; sample=bytearray(random.getrandbits(8) for _ in range({LENGTH}))"
        ).timeit()
        print(f"{f'Baseline time to execute md5:' : <30}{base:7.5f}")
        print()

        for algo in [x for x in sorted(list(hashlib.algorithms_guaranteed)) if x not in self.ALGO_IGNORED_LIST]:
            result = timeit.Timer(
                f"hashlib.{algo}(sample).hexdigest()",
                f"import hashlib; import random; sample=bytearray(random.getrandbits(8) for _ in range({LENGTH}))"
                ).timeit()
            print(f"{f'Time to execute {algo}:' : <30}{result:7.5f} | % diff {(base-result)/result:6.4f}%")

    def test__make_cache_key_Algorithm(self):
        """
        Benchmarks the _make_cache_key function with the default and all available configuration options

        Results:
            Baseline time to execute `core._make_cache_key`:       9.75964

            Time to execute `core._make_cache_key` using blake2b:  9.64733 | % diff 0.0116%
            Time to execute `core._make_cache_key` using blake2s:  9.55411 | % diff 0.0215%
            Time to execute `core._make_cache_key` using md5:      9.68687 | % diff 0.0075%
            Time to execute `core._make_cache_key` using sha1:     9.67266 | % diff 0.0090%
            Time to execute `core._make_cache_key` using sha224:   9.64335 | % diff 0.0121%
            Time to execute `core._make_cache_key` using sha256:   9.59824 | % diff 0.0168%
            Time to execute `core._make_cache_key` using sha384:   9.66517 | % diff 0.0098%
            Time to execute `core._make_cache_key` using sha3_224: 9.70930 | % diff 0.0052%
            Time to execute `core._make_cache_key` using sha3_256: 9.69792 | % diff 0.0064%
            Time to execute `core._make_cache_key` using sha3_384: 9.68906 | % diff 0.0073%
            Time to execute `core._make_cache_key` using sha3_512: 9.67534 | % diff 0.0087%
            Time to execute `core._make_cache_key` using sha512:   9.53044 | % diff 0.0240%
        """

        base = timeit.Timer(
            f'''class TimeitTestCore(TestCase):
                def timeit_test_baseline(self):
                    core._make_cache_key(
                        group=grp,
                        window=wnd,
                        rate=rte,
                        value=val,
                        methods="ALL",
                    )''',
            f"from django_ratelimit import core;"
            f"import random;"
            f"from string import ascii_letters;"
            f"from django.test import TestCase;"
            f"grp = ''.join(random.choices(ascii_letters, k={self.CACHE_KEY_SIZE}));"
            f"wnd = ''.join(random.choices(ascii_letters, k={self.CACHE_KEY_SIZE}));"
            f"rte = '/'.join([str(random.randint(1, 50)), random.choice(['s', 'm', 'h', 'd'])]);"
            f"val = ''.join(random.choices(ascii_letters, k={self.CACHE_KEY_SIZE}));"
        ).timeit()
        print(f"{f'Baseline time to execute `core._make_cache_key`:' : <55}{base:7.5f}")
        print()

        for algo in [x for x in sorted(list(hashlib.algorithms_guaranteed)) if x not in self.ALGO_IGNORED_LIST]:
            result = timeit.Timer(
                f"""class TimeitTestCore(TestCase):
                        def timeit_test_{algo}(self):
                            with self.settings(RATELIMIT_HASH_ALGORITHM='hashlib.{algo}'):
                                core._make_cache_key(
                                        group=grp,
                                        window=wnd,
                                        rate=rte,
                                        value=val,
                                        methods="ALL",
                                    )""",
                f"from django_ratelimit import core;"
                f"import random;"
                f"from string import ascii_letters;"
                f"from django.test import TestCase;"
                f"grp = ''.join(random.choices(ascii_letters, k={self.CACHE_KEY_SIZE}));"
                f"wnd = ''.join(random.choices(ascii_letters, k={self.CACHE_KEY_SIZE}));"
                f"rte = '/'.join([str(random.randint(1, 50)), random.choice(['s', 'm', 'h', 'd'])]);"
                f"val = ''.join(random.choices(ascii_letters, k={self.CACHE_KEY_SIZE}));"
            ).timeit()

            print(f"{f'Time to execute `core._make_cache_key` using {algo}:' : <55}{result:7.5f} | % diff {(base-result)/result:6.4f}%")
```

slick666 commented:
🤦‍♂️ I knew I was forgetting something. **Now** the default hash is sha256

jsocol reviewed (approved):
I like it, thank you!
